### PR TITLE
Bug 1894649: Update Kubelet service in-place instead of recreating

### PR DESCRIPTION
**Pull request by VaishnaviHire**
This commit ensures that the idempotency of kubelet service is
mainitained. Recreating the kubelet service caused the bug where
wmcb tried to access kubelet service that was marked for deletion.
The changes include removing deletion of kubelet service, and
ensuring that the kubelet service is created if it is not present,
else updating the existing one with custom specifications.

[Bug 1894649](https://issues.redhat.com/browse/OCPBUGSM-20185)

### CODE REVIEW

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:882]: The benefit of converting the args into a map is that if we want to change specific arguments that is made easier. In this case you are not doing that, and it seems like you are creating a new function to convert a []string into a map so that the `reconstructKubeletCmd` can be reused. While its good to reuse functions, this is adding unnecessary logic and obfuscating a simple operation. 

Why was the below solution we talked about unviable?
```	
        s := syscall.EscapeArg(kubeletpath)
	for _, v := range args {
		s += " " + syscall.EscapeArg(v)
	}
```
This code was borrowed from the [svc mgr library](https://github.com/golang/sys/blob/master/windows/svc/mgr/mgr.go#L126-L129)

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:182]: I think the approach taken here requires a commit/PR title change, something like `Update kubelet service in place instead of recreating`

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:455]: Given you are in the area, please update the comment

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:540]: This is a repeated pattern and should be modularized.

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:452]: This is more like `getInitialKubeletArgs()`.  You should update the comment stating that this function assumes `wmcb.kubeletArgs` have been populated. Ideally there should be a guard but I am not sure what that guard looks like.

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:527]: Do we need to call this after we finish configuration?

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:862]: If this is updating the kubelet cmd then why is it called `configureKubelet`? This should be `updateKubeletCmd`

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go:182]: Given this is a bug, I am not sure if the title should be same as in Bugzilla

File: test/e2e/bootstrapper_test.go
Comment (aravindhp) [inline test/e2e/bootstrapper_test.go:81]: Any reason to use 1 minute vs 30 seconds here?

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @VaishnaviHire. Please address my comments.

File: test/e2e/bootstrapper_test.go
Comment (VaishnaviHire) [reply test/e2e/bootstrapper_test.go:81]: At the start of the `Poll`, the kubelet service would be running, resulting in immediate return

File: pkg/bootstrapper/bootstrapper.go
Comment (mansikulkarni96) [inline pkg/bootstrapper/bootstrapper.go:186]: Correct me if I am wrong but isn't this function called on intialize kubelet as well and when the hybrid-overlay is not yet installed? Won't that always result in an error being returned here?

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [reply pkg/bootstrapper/bootstrapper.go:182]: I missed this. Please don't repeat the title of the bug in the PR. That is akin to saying you are introducing the bug. Mention **how** you are fixing the bug.

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go:527]: We are also calling [refresh](https://github.com/openshift/windows-machine-config-bootstrapper/blob/master/pkg/bootstrapper/kubelet.go#L116) after configuration. Do we still go ahead and call this function ?

Comment (aravindhp) [reply pkg/bootstrapper/bootstrapper.go:527]: I am going to defer to @sebsoto on this. I want to ensure we are closing all open handles.

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go:186]: You might be right. How do we handle errors returned by `OpenService` when the dependent services are installed ?

Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go:186]: we might have to check by ignoring a specific error here.

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go:527]: The refresh was important since we needed to close handles to the service manager in order to allow the service to be deleted. Now all we need to make sure to do is that we are closing it at the end. The process exiting should do that automatically, but to be safe we should explicitly disconnect

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go:882]: I agree that we can get away without converting the []string to map, and can update that.

However I am not sure why the above code snippet is required when we can pass `&c.BinaryPath` where c is the `inital config`

File: test/e2e/bootstrapper_test.go
Comment (VaishnaviHire) [reply test/e2e/bootstrapper_test.go:81]: Calling InitializeKubelet() on already running service involves stopping the service, updating the config and restarting the service, hence the increased buffer time

**Comment by LorbusChris** [comment]
nit: two typos in commit message:
```diff
-Bug 1894649: Idempotentcy of kubelet service is broken
+Bug 1894649: Idempotency of kubelet service is broken

-This commit ensures that the idempotentcy of kubelet service is
+This commit ensures that the idempotency of kubelet service is
-mainitained.The changes include ensuring that the kubelet service is
+maintained.The changes include ensuring that the kubelet service is
created if it is not present, else updating the existing kubelet
service with custom specifications.
```

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go:452]: We do have some guards [here](https://github.com/openshift/windows-machine-config-bootstrapper/pull/239/files#diff-e9ddf87a450b985b61b945593604139617a74d4dbacff8b018acfe18c5880c2cR483). Show we return error if the key doesn`t exist ?

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [reply pkg/bootstrapper/bootstrapper.go:527]: Never assume exiting will clean things up automatically, especially on Windows ;-)

File: test/e2e/bootstrapper_test.go
Comment (aravindhp) [reply test/e2e/bootstrapper_test.go:81]: Not a fan of such sleeps. I would prefer we wait for the expected result with a timeout.

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [reply pkg/bootstrapper/bootstrapper.go:452]: That is not a guard. A guard would something that returns an error if the function is called before the args have been populated. 

>Should we return error if the key doesn`t exist ?

Not sure if we can do that. Some of the args are optional.

File: pkg/bootstrapper/bootstrapper.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go:186]: sorry I missed this, but this is only called when kubelet service is already running and not when we are creating it for the first time. Thinking of this scenario I think its fine checking for any error here.

File: pkg/bootstrapper/bootstrapper.go
Comment (mansikulkarni96) [inline pkg/bootstrapper/bootstrapper.go:548]: comment needs to be updated

**Review by mansikulkarni96 — COMMENTED** [review]
Thanks for fixing this @VaishnaviHire PTAL ta my comments.

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go]: Its important to escape the arguments in case we are sourcing any problematic strings from the ignition file

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go:527]: Won`t the refresh function be required here since it calls, [svcobj.UpdateConfig](https://github.com/openshift/windows-machine-config-bootstrapper/blob/master/pkg/bootstrapper/kubelet.go#L121). We are also calling [Disconnect](https://github.com/openshift/windows-machine-config-bootstrapper/blob/master/cmd/bootstrapper/initialize_kubelet.go#L72) after IntializeKubelet, that should clean things up

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go]: Discussed this, neither of the input is needed, we could generate the string and update the c.BinaryPath

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go:185]: Ignoring `service does not exist` 

> we might have to check by ignoring a specific error here.

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:566]: Why is this separation being done?

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:573]: This shouldnt be necessary

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:876]: extra space

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go:573]: Wouldn`t we need to call [updateConfig](https://github.com/openshift/windows-machine-config-bootstrapper/blob/master/pkg/bootstrapper/kubelet.go#L121) ?

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go:566]: Thanks for catching that. I need to revert it. Added it as part of a debug.

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go:573]: sorry you are right, read it too quick and thought that was refreshing the manager

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:557]: Why do you need a space before the command?

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:597]: s/created/present

File: pkg/bootstrapper/kubelet.go
Comment (aravindhp) [inline pkg/bootstrapper/kubelet.go:130]: Such times should be expressed as constants.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go:492]: Yes, we pass it as exepath [here](https://github.com/openshift/windows-machine-config-bootstrapper/blob/7720182cdc978c38df06eed22cf36abbf7c9ee0a/vendor/golang.org/x/sys/windows/svc/mgr/mgr.go#L116)

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [reply pkg/bootstrapper/bootstrapper.go:527]: Yes. [ConfigureCNICmd](https://github.com/openshift/windows-machine-config-bootstrapper/blob/master/cmd/bootstrapper/configure_cni.go#L72)

File: pkg/bootstrapper/kubelet.go
Comment (aravindhp) [inline pkg/bootstrapper/kubelet.go:14]: This is not the wait time but rather the interval at which you will poll the service. `svcPollInterval`. Given it is is in `kubelet.go` no reason to have `kubelet`.

File: pkg/bootstrapper/kubelet.go
Comment (aravindhp) [inline pkg/bootstrapper/kubelet.go:16]: Similarly I suggest using `svcRunTimeout`. Please set comment correctly too.

File: test/e2e/bootstrapper_test.go
Comment (aravindhp) [inline test/e2e/bootstrapper_test.go:87]: Please add a constant for `2*time.Minute`. `waitTimeKubeletLog`.

File: test/e2e/bootstrapper_test.go
Comment (aravindhp) [inline test/e2e/bootstrapper_test.go:27]: `pollIntervalKubeletLog`

**Review by aravindhp — APPROVED** [review]
/approve

with the caveat that the variable names are fixed up.

File: pkg/bootstrapper/bootstrapper.go
Comment (LorbusChris) [inline pkg/bootstrapper/bootstrapper.go:876]: tiny nit:
```suggestion
			return nil, fmt.Errorf("error getting dependent services for kubelet %v", err)
```

**Review by LorbusChris — APPROVED** [review]
/lgtm

**Comment by LorbusChris** [comment]
@VaishnaviHire the nit above, plus the ones mentioned in https://github.com/openshift/windows-machine-config-bootstrapper/pull/239#issuecomment-737392572 - not blocking on it, though :)

**Pull request merged**
